[PyTorch Distributed] Add debug hint for NCCL async system error

kwen2501 opened the pull request:
Summary:
add a debug hint that async system error can be caused by unexpected exit of
a remote process if not an actual network issue. For example, the exit of the remote process
can cause a closed network connection error at a local process. The hint helps to direct
the debug focus to the remote process.

Test Plan: unit tests

Differential Revision: D34702348

pritamdamania87 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -970,9 +970,16 @@ std::exception_ptr ProcessGroupNCCL::checkForNCCLErrorsInternal(
     if (ncclAsyncErr != ncclSuccess) {
+      std::string additionalMsg;
+      if (ncclAsyncErr == ncclSystemError) {
+        additionalMsg = c10::str("\n",
```
Probably better to extend the error message directly here: https://github.com/pytorch/pytorch/blob/master/torch%2Fcsrc%2Fdistributed%2Fc10d%2FNCCLUtils.hpp#L28

kwen2501 replied:
Appreciate your comment @pritamdamania87 ! 
I put the additional message here because I need to rely on the fact that this is an **asynchronous** NCCL error (i.e. dynamic during a NCCL operation), rather than applying the hint to all ncclSystemError's which may also include those immediately returned during NCCL initialization. Errors like "connection closed by remote peer" most happen asynchronously during NCCL execution.

kwen2501 replied:
Updated code change to reflect @pritamdamania87 's comment.

rohan-varma reviewed (approved):
LGTM given @kwen2501 's reasoning on the response to Pritam's comment. Although let's wait for @pritamdamania87 to confirm.

pritamdamania87 commented on torch/csrc/distributed/c10d/NCCLUtils.hpp:
```diff
@@ -25,7 +25,8 @@ const inline char* getNcclErrorDetailStr(ncclResult_t error, c10::optional<std::
     case ncclSystemError:
-      return "ncclSystemError: System call (socket, malloc, munmap, etc) failed.";
+      return "ncclSystemError: System call (e.g. socket, malloc) or external library call (e.g. ibverbs) failed or device error. "
+        "It can be also caused by unexpected exit of a remote peer, check NCCL warnings for peer id.";
```
Should we mention this instead:

```
It can be also caused by unexpected exit of a remote peer, please check NCCL logs (after enabling NCCL_DEBUG=WARN or INFO) for the exact reason of the failure.
```

pritamdamania87 commented on torch/csrc/distributed/c10d/NCCLUtils.hpp:
```diff
@@ -25,7 +25,8 @@ const inline char* getNcclErrorDetailStr(ncclResult_t error, c10::optional<std::
+      return "ncclSystemError: System call (e.g. socket, malloc) or external library call (e.g. ibverbs) failed or device error. "
+        "It can be also caused by unexpected exit of a remote peer, check NCCL warnings for peer id.";
     case ncclInternalError:
       return "ncclInternalError: Internal check failed. This is either a bug in NCCL or due to memory corruption";
```
Not for this PR, but in a separate PR maybe we should add the line "Please check NCCL logs (after enabling NCCL_DEBUG=WARN or INFO) for the exact reason of the failure." to all other messages too.

kwen2501 committed:
[PyTorch Distributed] Add debug hint for NCCL async system error (#73897)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/73897

add a debug hint that async system error can be caused by unexpected exit of
a remote process if not an actual network issue. For example, the exit of the remote process
can cause a closed network connection error at a local process. The hint helps to direct
the debug focus to the remote process.

Test Plan: unit tests

Reviewed By: pritamdamania87, rohan-varma

Differential Revision: D34702348

fbshipit-source-id: 85ccebc25e4c3a685dfb7d2bc2d981778cd08cd7
